### PR TITLE
Improve TopK CUDA execution path while sampling and when vocab size is sufficiently large

### PR DESCRIPTION
**Pull request by hariharans29**
The 20B GPT OSS model has a vocab size of 200k+ and while sampling (with no batching), the top_k sampling boils down to finding the top_k elements in a large array (batch_size == 1) which is a hard-problem to parallelize and optimally use hardware resources on modern GPUs. 

A profile of the 20B GPT OSS model on A100 with default top_k parameter (top_k = 50) shows the top_k kernel as a huge bottleneck (about 50%) of the forward pass as can be seen from this nsys snapshot:

image

As can be seen from the kernel launch parameters, only one threadblock (since batch_size = 1 in sampling) is used and 256 threads are utilized to search top_k elements from an array of about 200k+elements

image

This leads to severe under-utilization of the hardware.

To mitigate the problem, thje following were tried:

1) Do distributed "top_k" search:
    Use multiple threadblocks to each individually pick top_k elements from their own individual shards of the array and do a final "reduction" tro pick the final top_k elements. This gave about 5% speed-up. The bottleneck was the reduction step.

2) Use sort and pick top_k
    For sufficiently large arrays, sorting using CUB's device fragmented radix sort (which has clever bit manipulation tricks to perform radix sort for `floats` and `halfs`  uses the hardware at hand much more optimally and outperforms any tradional top_k picking algorithms on device. See https://github.com/pytorch/pytorch/issues/22812 and https://github.com/pytorch/pytorch/pull/68632.

Hence, the final approach is to go with option 2 based on a tunable heuristic.

Here are the before and after sampling TPS numbers for 20B GPT OSS on A100 for this optimization for the same prompt (for top_k = 50 which seems to be FoundryLocal's default config) - about **35%** more sampling throughput
 
image

### CODE REVIEW

File: src/cuda/cuda_sampling.cu [hunk 1]
```diff
@@ -548,15 +548,26 @@ void GetTopKSubset(SamplingData* data, cudaStream_t stream, float* scores_in, fl
...
 
-  if (k <= 4) {
+  bool use_sort_and_pick_topk = false;
+
+  // TODO(hasesh): Tweak this heuristic by performing more benchmarks.
+  // It is likely that the sort and pick top_k approach works well for
+  // smallish batch_sizes on most hardware. For now, we limit it to
+  // sampling (batch_size == 1) and when vocab_size is "pretty large"
+  // (also needs benchmarking to identify cut-off).
+  if (batch_size == 1 && vocab_size >= 100000) {
```
Comment (kunal-vaishnavi) [inline src/cuda/cuda_sampling.cu]: Should we make this heuristic accessible outside of the build to allow runtime tuning (e.g. as an env variable)?

File: src/cuda/cuda_sampling.cu [hunk 2]
```diff
@@ -548,15 +548,26 @@ void GetTopKSubset(SamplingData* data, cudaStream_t stream, float* scores_in, fl
                              k,                          \
                              temperature);
 
-  if (k <= 4) {
+  bool use_sort_and_pick_topk = false;
```
Comment (jiafatom) [inline src/cuda/cuda_sampling.cu:577]: How does this change affect model accuracy in general? It seems changing sampling algorithm? Then we need to measure it on babelbench to make sure the model accuracy does not get affected. Need more information from author to proceed.

Comment (hariharans29) [reply src/cuda/cuda_sampling.cu:577]: Could you elaborate on how it changes the sampling algo ? It just changes the way in which the top_k elements are retrieved. Happy to run it through any quality benchmarks though.

File: src/cuda/cuda_sampling.cu (hunk 1 above)
Comment (hariharans29) [reply src/cuda/cuda_sampling.cu]: Personally I think that is overkill for now.

Ultimately, the heuristic will be something like - 

`(batch_size <= some_batch_threshold && vocab_size >= some_vocab_size_threshold)`

 I think, ultimately, these can be fine-tuned within the build as we keep profiling models ?

**Review by kunal-vaishnavi — APPROVED** [review]

**Comment by tianleiwu** [comment]
A partial sort algorithm like cub::DeviceSelect is generally more efficient than a full sort. 

Example code snippet generated by LLM:
```
#include <cub/cub.cuh>

// A simple kernel to partition the original scores and indices based on a threshold
__global__ void PartitionTopKKernel(const float* scores_in, const int* indices_in,
                                    float* scores_out, int* indices_out,
                                    int vocab_size, float threshold, int k) {
    // Use an atomic counter to track the number of elements written to the output
    __shared__ int output_idx_counter;
    if (threadIdx.x == 0) {
        output_idx_counter = 0;
    }
    __syncthreads();

    for (int i = threadIdx.x; i < vocab_size; i += blockDim.x) {
        if (scores_in[i] >= threshold) {
            int write_idx = atomicAdd(&output_idx_counter, 1);
            // Ensure we don't write more than k elements, which can happen with duplicate scores
            if (write_idx < k) {
                scores_out[write_idx] = scores_in[i];
                indices_out[write_idx] = indices_in[i];
            }
        }
    }
}

// A new version of GetTopKSubset using partial sort for the general case
void GetTopKSubset_PartialSort(SamplingData* data, cudaStream_t stream, float* scores_in, float* scores_out, int* indices_out, int vocab_size, int batch_size, int k, float temperature) {
    // This implementation is simplified for batch_size = 1 as per the use case.
    // Extending to batch_size > 1 would require cub::DeviceSegmentedSelect.
    if (batch_size != 1) {
        // Fallback or error for simplicity.
        // The original LaunchSort using CUB Segmented Radix Sort is a good general solution.
        LaunchSort(data, stream, scores_in, data->scores_buffer.get(), indices_out, vocab_size, batch_size);
        DispatchBlockwiseSoftmaxForward<false>(stream, scores_out, const_cast<const float*>(data->scores_buffer.get()), k, vocab_size, k, batch_size);
        return;
    }

    // --- Step 1: Find the value of the k-th largest element ---
    // We want the k-th largest, which is the (vocab_size - k)-th smallest element.
    size_t temp_storage_bytes = 0;
    int rank = vocab_size - k;
    
    // Get temporary storage size
    cub::DeviceSelect::Select(nullptr, temp_storage_bytes, scores_in, data->scores_buffer.get(), vocab_size, &rank);
    data->temp_storage_bytes = temp_storage_bytes; // Assuming temp_buffer can be resized or is large enough
    
    // Run the selection
    cub::DeviceSelect::Select(data->temp_buffer.get(), data->temp_storage_bytes, scores_in, data->scores_buffer.get(), vocab_size, &rank);
    
    // data->scores_buffer now contains a single float: the k-th largest score.
    float threshold = 0.0f;
    cudaMemcpyAsync(&threshold, data->scores_buffer.get(), sizeof(float), cudaMemcpyDeviceToHost, stream);
    cudaStreamSynchronize(stream); // Sync to get the threshold value on the CPU.
                                   // A device-side alternative is preferred for max performance.

    // --- Step 2: Partition the full arrays based on the threshold ---
    // We need original indices [0, 1, 2, ..., vocab_size-1] to partition alongside the scores.
    LaunchPopulateIndices(data->indices_in.get(), vocab_size, 1, stream);
    
    // This custom kernel filters both scores and indices into a compact list of size k.
    // The output goes into the first k slots of scores_buffer and indices_out.
    dim3 grid(1);
    dim3 block(1024);
    PartitionTopKKernel<<<grid, block, 0, stream>>>(
        scores_in, data->indices_in.get(),
        data->scores_buffer.get(), indices_out,
        vocab_size, threshold, k);

    // --- Step 3: Apply softmax to the k selected elements ---
    // The input stride is now k, as PartitionTopKKernel created a compact list.
    DispatchBlockwiseSoftmaxForward<false>(stream, scores_out, const_cast<const float*>(data->scores_buffer.get()), k, k, k, batch_size);
}
```

**Review by tianleiwu — APPROVED** [review]

File: src/cuda/cuda_sampling.cu [hunk 3]
```diff
@@ -548,15 +548,26 @@ void GetTopKSubset(SamplingData* data, cudaStream_t stream, float* scores_in, fl
...
+  } else if (!use_sort_and_pick_topk && k <= 8) {
     GetTopK(8);
-  } else if (k <= 16) {
+  } else if (!use_sort_and_pick_topk && k <= 16) {
     GetTopK(16);
-  } else if (k <= 32) {
+  } else if (!use_sort_and_pick_topk && k <= 32) {
     GetTopK(32);
-  } else if (k <= 64) {
+  } else if (!use_sort_and_pick_topk && k <= 64) {
```
Comment (jiafatom) [inline src/cuda/cuda_sampling.cu:596]: For `GetTopK`, there is a `scores_out[i + batch * k] = top_k_sequence.value[i] / temperature;` in 
 https://github.com/microsoft/onnxruntime-genai/blob/b6d74e2e7efe2ae4d2b4eda7a10bcbd2eb3b1ff3/src/cuda/cuda_sampling.cu#L469, seems we need change in `LaunchSort`

Comment (hariharans29) [reply src/cuda/cuda_sampling.cu:596]: Good catch, thanks - seems like the temperature is not honored in the sort based approach. 
Added the temperature scaling logic to the "sort + top_k" path

Comment (hariharans29) [reply src/cuda/cuda_sampling.cu:596]: Are there any temperature specific tests I can add to cover this case ?

File: src/cuda/cuda_sampling.cu [hunk 4]
```diff
@@ -528,13 +547,20 @@ void LaunchSampleKernel(SamplingData* data, cudaStream_t stream, float* scores,
...
   LaunchPopulateIndices(data->indices_in.get(), vocab_size, batch_size, stream);
   LaunchSortPairs<float>(data->temp_buffer.get(), data->temp_storage_bytes, scores_in, scores_out,
                          data->indices_in.get(), indices_out, vocab_size * batch_size, batch_size, data->offsets.get(),
                          stream, /*is_descending*/ true);
+
+  if (temperature != 1.f) {
+    // TODO: Fuse the temperature scaling with subsequent Softmax kernel - currently it isn't a very costly
+    // kernel, so keep it unfused
+    // We only need to scale the "top_k" elements from the sorted array (which is the output of the above kernel)
+    LaunchTemperatureScaling(scores_out, temperature, k, stream);
```
Comment (jiafatom) [inline src/cuda/cuda_sampling.cu]: Can we have same param order to match with `LaunchSort`? like `score_out, stream, k, temperature`?

Comment (hariharans29) [reply src/cuda/cuda_sampling.cu]: Sure, fixed the ordering
